File: InverceMod.Logging/src/main/java/com/inverce/logging/Log.java
package com.inverce.logging;

import android.content.Context;
import android.support.annotation.StringRes;

@SuppressWarnings("unused")
public class Log {
    private static boolean DEBUG_MODE = true;
    private static final String BASE_TAG = "||";
    private static final String libraryPackage = "com.empik.lib";
    private static       String ApplicationPackage = "com.example.exampleapp";

    public static final int VERBOSE = 2, DEBUG = 3, INFO = 4, WARN = 5, ERROR = 6;
    public static final int EXCEPTION = 7, ASSERT = 8, NONE = Integer.MAX_VALUE;
    private static final int FULL = 1, SIMPLER = 2, SIMPLEST = 3;

    private static int LOGGING_LEVEL = VERBOSE;

    private static LogListener listener;
    private static Context ApplicationContext;

    public static void setListener(LogListener listener) {
        Log.listener = listener;
    }

    public static void setLogLevel(int LOGGING_LEVEL) {
        Log.LOGGING_LEVEL = LOGGING_LEVEL;
    }

    public static void setDebugMode(boolean debugMode) {
        DEBUG_MODE = debugMode;
    }

    public static void init(Context context) {
        ApplicationPackage = context.getPackageName();
        ApplicationContext = context.getApplicationContext();
    }

    private static Context context() {
        return ApplicationContext;
    }

    public static void handleMsg(int lvl, int tag, int msg, Object ... o) {
        if (!(DEBUG_MODE && LOGGING_LEVEL <= lvl)) {
            return;
        }

        if (tag == -1) {
            handleMsg(lvl, null, context().getString(msg, o));
        } else {
            handleMsg(lvl, context().getString(tag), context().getString(msg, o));
        }
    }

    public static void handleMsg(int lvl, String tag, String msg, Object ... o) {
        if (!(DEBUG_MODE && LOGGING_LEVEL <= lvl && msg != null)) {
            return;
        }

        tag = (tag != null) ? (BASE_TAG + "." + tag) : BASE_TAG;
        msg = (o.length > 0) ? String.format(msg, o) : msg;

        if (listener != null && lvl < EXCEPTION) {
            if (listener.handleMsg(lvl, tag, msg)) {
                return;
            }
        }
        switch (lvl) {
            case VERBOSE : android.util.Log.v(tag, msg); break;
            case DEBUG   : android.util.Log.d(tag, msg); break;
            case INFO    : android.util.Log.i(tag, msg); break;
            case WARN    : android.util.Log.w(tag, msg); break;
            case ERROR   : android.util.Log.e(tag, msg); break;
            case ASSERT  : handleExc(ASSERT, tag, msg, new AssertionError(msg)); break;
        }
    }

    public static void handleExc(int lvl, int tag, int msg, Throwable o) {
        if (!(DEBUG_MODE && LOGGING_LEVEL <= lvl)) {
            return;
        }

        if (tag == -1) {
            handleExc(lvl, null, context().getString(msg), o);
        } else {
            handleExc(lvl, context().getString(tag), context().getString(msg), o);
        }
    }

    public static void handleExc(int simple_lvl, String tag, String msg, Throwable o) {
        if (!(DEBUG_MODE && LOGGING_LEVEL <= EXCEPTION)) {
            return;
        }

        if (listener != null) {
            listener.handleExc(simple_lvl, tag, msg, o);
        }

        tag = (tag != null) ? (BASE_TAG + "." + tag) : BASE_TAG;
        msg = (msg != null ? msg + ": " : "" ) + "<" + o.getClass().getSimpleName() + "> "+ o.getMessage();

        android.util.Log.e(tag, msg);

        switch (simple_lvl) {
            case FULL: o.printStackTrace(); break;
            case SIMPLER:
            default:
                android.util.Log.w("", msg);
                StackTraceElement[] stackTrace = o.getStackTrace();
                for (StackTraceElement aStackTrace : stackTrace) {
                    String className = aStackTrace.getClassName();
                    if (className.contains(libraryPackage) || className.contains(ApplicationPackage)) {
                        android.util.Log.w(tag, aStackTrace.toString());
                    }
                }

                if (o.getCause() != null)
                    android.util.Log.w(tag, "Caused by: " + "<" + o.getCause().getClass().getSimpleName() + "> "+ o.getCause().getMessage());
                break;
        }
    }

    public static void v(String tag, String message, Object ... o) {                                handleMsg(VERBOSE,      tag,    message,  o); }
    public static void v(String tag, String message) {                                              handleMsg(VERBOSE,      tag,    message    ); }
    public static void v(String message) {                                                          handleMsg(VERBOSE,      null,   message    ); }
    public static void v(String message, Object ... o) {                                            handleMsg(VERBOSE,      null,   message,  o); }
    public static void d(String tag, String message, Object ... o) {                                handleMsg(DEBUG,        tag,    message,  o); }
    public static void d(String tag, String message) {                                              handleMsg(DEBUG,        tag,    message    ); }
    public static void d(String message) {                                                          handleMsg(DEBUG,        null,   message    ); }
    public static void d(String message, Object ... o) {                                            handleMsg(DEBUG,        null,   message,  o); }
    public static void i(String tag, String message, Object ... o) {                                handleMsg(INFO, tag, message, o); }
    public static void i(String tag, String message) {                                              handleMsg(INFO,         tag,    message    ); }
    public static void i(String message) {                                                          handleMsg(INFO,         null,   message    ); }
    public static void i(String message, Object ... o) {                                            handleMsg(INFO,         null,   message,  o); }
    public static void w(String tag, String message, Object ... o) {                                handleMsg(WARN,         tag,    message,  o); }
    public static void w(String tag, String message) {                                              handleMsg(WARN,         tag,    message    ); }
    public static void w(String message) {                                                          handleMsg(WARN,         null,   message    ); }
    public static void w(String message, Object ... o) {                                            handleMsg(WARN,         null,   message,  o); }
    public static void e(String tag, String message, Object ... o) {                                handleMsg(ERROR,        tag,    message,  o); }
    public static void e(String tag, String message) {                                              handleMsg(ERROR,        tag,    message    ); }
    public static void e(String message) {                                                          handleMsg(ERROR,        null,   message    ); }
    public static void e(String message, Object ... o) {                                            handleMsg(ERROR,        null,   message,  o); }
    public static void a(String tag, String message, Object ... o) {                                handleMsg(ASSERT,       tag,    message,  o); }
    public static void a(String tag, String message) {                                              handleMsg(ASSERT,       tag,    message    ); }
    public static void a(String message) {                                                          handleMsg(ASSERT,       null,   message    ); }
    public static void a(String message, Object ... o) {                                            handleMsg(ASSERT,       null,   message,  o); }

    public static void v(@StringRes int tag, @StringRes int message, Object ... o) {                handleMsg(VERBOSE,      tag,    message,  o); }
    public static void v(@StringRes int tag, @StringRes int message) {                              handleMsg(VERBOSE,      tag,    message    ); }
    public static void v(@StringRes int message) {                                                  handleMsg(VERBOSE,      -1,     message    ); }
    public static void v(@StringRes int message, Object ... o) {                                    handleMsg(VERBOSE,      -1,     message,  o); }
    public static void d(@StringRes int tag, @StringRes int message, Object ... o) {                handleMsg(DEBUG,        tag,    message,  o); }
    public static void d(@StringRes int tag, @StringRes int message) {                              handleMsg(DEBUG,        tag,    message    ); }
    public static void d(@StringRes int message) {                                                  handleMsg(DEBUG,        -1,     message    ); }
    public static void d(@StringRes int message, Object ... o) {                                    handleMsg(DEBUG,        -1,     message,  o); }
    public static void i(@StringRes int tag, @StringRes int message, Object ... o) {                handleMsg(INFO,         tag,    message,  o); }
    public static void i(@StringRes int tag, @StringRes int message) {                              handleMsg(INFO,         tag,    message    ); }
    public static void i(@StringRes int message) {                                                  handleMsg(INFO,         -1,     message    ); }
    public static void i(@StringRes int message, Object ... o) {                                    handleMsg(INFO,         -1,     message,  o); }
    public static void w(@StringRes int tag, @StringRes int message, Object ... o) {                handleMsg(WARN,         tag,    message,  o); }
    public static void w(@StringRes int tag, @StringRes int message) {                              handleMsg(WARN,         tag,    message    ); }
    public static void w(@StringRes int message) {                                                  handleMsg(WARN,         -1,     message    ); }
    public static void w(@StringRes int message, Object ... o) {                                    handleMsg(WARN,         -1,     message,  o); }
    public static void e(@StringRes int tag, @StringRes int message, Object ... o) {                handleMsg(ERROR,        tag,    message,  o); }
    public static void e(@StringRes int tag, @StringRes int message) {                              handleMsg(ERROR,        tag,    message    ); }
    public static void e(@StringRes int message) {                                                  handleMsg(ERROR,        -1,     message    ); }
    public static void e(@StringRes int message, Object ... o) {                                    handleMsg(ERROR,        -1,     message,  o); }
    public static void a(@StringRes int tag, @StringRes int message, Object ... o) {                handleMsg(ASSERT,       tag,    message,  o); }
    public static void a(@StringRes int tag, @StringRes int message) {                              handleMsg(ASSERT, tag, message); }
    public static void a(@StringRes int message) {                                                  handleMsg(ASSERT, -1, message); }
    public static void a(@StringRes int message, Object ... o) {                                    handleMsg(ASSERT, -1, message, o); }

    public static void ex(Throwable o) {                                                            handleExc(FULL,         null, null, o); }
    public static void exs(Throwable o) {                                                           handleExc(SIMPLER,      null, null, o); }
    public static void exm(Throwable o) {                                                           handleExc(SIMPLEST,     null, null, o); }

    public static void ex(String tag, String message, Throwable o) {                                handleExc(FULL,         tag,  message, o); }
    public static void ex(String message, Throwable o) {                                            handleExc(FULL,         null, message, o); }
    public static void exs(String tag, String message, Throwable o) {                               handleExc(SIMPLER,      tag,  message, o); }
    public static void exs(String message, Throwable o) {                                           handleExc(SIMPLER,      null, message, o); }
    public static void exm(String tag, String message, Throwable o) {                               handleExc(SIMPLEST,     tag,  message, o); }
    public static void exm(String message, Throwable o) {                                           handleExc(SIMPLEST,     null, message, o); }

    public static void ex(@StringRes int tag, @StringRes int message, Throwable o) {                handleExc(FULL,         tag,  message, o); }
    public static void ex(@StringRes int message, Throwable o) {                                    handleExc(FULL,         -1,   message, o); }
    public static void exs(@StringRes int tag, @StringRes int message, Throwable o) {               handleExc(SIMPLER,      tag,  message, o); }
    public static void exs(@StringRes int message, Throwable o) {                                   handleExc(SIMPLER,      -1,   message, o); }
    public static void exm(@StringRes int tag, @StringRes int message, Throwable o) {               handleExc(SIMPLEST,     tag,  message, o); }
    public static void exm(@StringRes int message, Throwable o) {                                   handleExc(SIMPLEST,     -1,   message, o); }

}

File: Core/src/main/java/com/inverce/mod/v2/core/Log.kt
package com.inverce.mod.v2.core

import android.support.annotation.StringRes
import com.inverce.mod.v2.core.LoggerInternal.Companion.printLocalCauseByStack
import com.inverce.mod.v2.core.LoggerInternal.Companion.printLocalExceptionStack
import com.inverce.mod.v2.core.LoggerInternal.Companion.str
import java.util.*

/**
 * Wrapper around android Log class, with additional benefits.
 */
object Log {
    val logger: Logger = Logger().apply {
        logLevel = VERBOSE
        enable = true
        listener = null
    }

    const val VERBOSE = Logger.VERBOSE
    const val DEBUG = Logger.DEBUG
    const val INFO = Logger.INFO
    const val WARN = Logger.WARN
    const val ERROR = Logger.ERROR
    const val EXCEPTION = Logger.EXCEPTION
    const val ASSERT = Logger.ASSERT
    const val NONE = Logger.NONE

    const val EX_FULL = Logger.EX_FULL
    const val EX_SIMPLER = Logger.EX_SIMPLER
    const val EX_SIMPLEST = Logger.EX_SIMPLEST

    var logLevel: Int?
        get() = logger.logLevel
        set(value) { logger.logLevel = value }

    var enable: Boolean?
        get() = logger.enable
        set(value) { logger.enable = value }

    var listener: LogListener?
        get() = logger.listener
        set(value) { logger.listener = value }


    @JvmOverloads @JvmStatic fun v(message: String?, vararg o: Any = emptyArray(), tag: String? = null)  = logger.handleMsg(VERBOSE, tag, message, o)
    @JvmOverloads @JvmStatic fun d(message: String?, vararg o: Any = emptyArray(), tag: String? = null)  = logger.handleMsg(DEBUG, tag, message, o)
    @JvmOverloads @JvmStatic fun i(message: String?, vararg o: Any = emptyArray(), tag: String? = null)  = logger.handleMsg(INFO, tag, message, o)
    @JvmOverloads @JvmStatic fun w(message: String?, vararg o: Any = emptyArray(), tag: String? = null)  = logger.handleMsg(WARN, tag, message, o)
    @JvmOverloads @JvmStatic fun e(message: String?, vararg o: Any = emptyArray(), tag: String? = null)  = logger.handleMsg(ERROR, tag, message, o)
    @JvmOverloads @JvmStatic fun a(message: String?, vararg o: Any = emptyArray(), tag: String? = null)  = logger.handleMsg(ASSERT, tag, message, o)

    @JvmOverloads @JvmStatic fun v(@StringRes message: Int, vararg o: Any = emptyArray())  = logger.handleMsg(VERBOSE,  null, str(message), o)
    @JvmOverloads @JvmStatic fun d(@StringRes message: Int, vararg o: Any = emptyArray())  = logger.handleMsg(DEBUG,    null, str(message), o)
    @JvmOverloads @JvmStatic fun i(@StringRes message: Int, vararg o: Any = emptyArray())  = logger.handleMsg(INFO,     null, str(message), o)
    @JvmOverloads @JvmStatic fun w(@StringRes message: Int, vararg o: Any = emptyArray())  = logger.handleMsg(WARN,     null, str(message), o)
    @JvmOverloads @JvmStatic fun e(@StringRes message: Int, vararg o: Any = emptyArray())  = logger.handleMsg(ERROR,    null, str(message), o)
    @JvmOverloads @JvmStatic fun a(@StringRes message: Int, vararg o: Any = emptyArray())  = logger.handleMsg(ASSERT,   null, str(message), o)

    @JvmOverloads @JvmStatic fun ex (t: Throwable, message: String? = null, vararg o: Any = emptyArray(), tag: String? = null) = logger.handleExc(EX_FULL, t, tag, message, o)
    @JvmOverloads @JvmStatic fun exs(t: Throwable, message: String? = null, vararg o: Any = emptyArray(), tag: String? = null) = logger.handleExc(EX_SIMPLER, t, tag, message, o)
    @JvmOverloads @JvmStatic fun exm(t: Throwable, message: String? = null, vararg o: Any = emptyArray(), tag: String? = null) = logger.handleExc(EX_SIMPLEST, t, tag, message, o)
    @JvmOverloads @JvmStatic fun ex (t: Throwable, @StringRes message: Int, vararg o: Any = emptyArray(), @StringRes tag: Int = 0) = logger.handleExc(EX_FULL, t, str(tag), str(message), o)
    @JvmOverloads @JvmStatic fun exs(t: Throwable, @StringRes message: Int, vararg o: Any = emptyArray(), @StringRes tag: Int = 0) = logger.handleExc(EX_SIMPLER, t, str(tag), str(message), o)
    @JvmOverloads @JvmStatic fun exm(t: Throwable, @StringRes message: Int, vararg o: Any = emptyArray(), @StringRes tag: Int = 0) = logger.handleExc(EX_SIMPLEST, t, str(tag), str(message), o)
}

open class Logger(protected val baseTag: String = "|>") {
    companion object {
        const val VERBOSE = 2
        const val DEBUG = 3
        const val INFO = 4
        const val WARN = 5
        const val ERROR = 6
        const val EXCEPTION = 7
        const val ASSERT = 8
        const val NONE = Integer.MAX_VALUE

        const val EX_FULL = -1
        const val EX_SIMPLER = -2
        const val EX_SIMPLEST = -3
    }

    /**
     * If null use Log defaults
     */
    var logLevel: Int? = null
    var enable: Boolean? = null
    var listener: LogListener? = null

    private inline fun <T> pVal(pVal: (Logger) -> T?): T? = when {
        this == Log.logger -> pVal(this)
        else -> pVal(this) ?: pVal(Log.logger)
    }

    protected fun dispatchMessage(lvl: Int, tag: String, msg: String) {
        when (lvl) {
            VERBOSE -> android.util.Log.v(tag, msg)
            DEBUG -> android.util.Log.d(tag, msg)
            INFO -> android.util.Log.i(tag, msg)
            WARN -> android.util.Log.w(tag, msg)
            ERROR -> android.util.Log.e(tag, msg)
            EXCEPTION, ASSERT -> handleExc(ASSERT, AssertionError(msg), tag, msg, emptyArray())
        }
    }

    fun shouldLog(lvl: Int) = (pVal { enable } ?: false) && (pVal { logLevel } ?: VERBOSE) <= lvl

    fun handleMsg(lvl: Int, tag: String?, msg: String?, o: Array<out Any>) {
        if (!shouldLog(lvl) || msg == null) return
        if (pVal { listener }?.handleMsg(lvl, tag, msg) == true) return
        dispatchMessage(lvl, "$baseTag${tag ?: ""}", String.format(Locale.getDefault(), msg, o))
    }

    fun handleExc(simple_lvl: Int, t: Throwable, tag: String?, msg: String?, o: Array<out Any>) {
        if (pVal { listener }?.handleExc(simple_lvl, t, tag, msg, o) == true) return
        android.util.Log.e("$baseTag${tag ?: ""}", String.format(Locale.getDefault(), msg ?: "$t", o))
        val pack = context.packageName
        when (simple_lvl) {
            EX_FULL -> t.printStackTrace()
            EX_SIMPLER -> printLocalExceptionStack(tag, pack, t)
            EX_SIMPLEST -> printLocalCauseByStack(tag, 2, t)
        }
    }

    @JvmOverloads fun v(message: String?, vararg o: Any = emptyArray(), tag: String? = null)  = handleMsg(VERBOSE, tag, message, o)
    @JvmOverloads fun d(message: String?, vararg o: Any = emptyArray(), tag: String? = null)  = handleMsg(DEBUG, tag, message, o)
    @JvmOverloads fun i(message: String?, vararg o: Any = emptyArray(), tag: String? = null)  = handleMsg(INFO, tag, message, o)
    @JvmOverloads fun w(message: String?, vararg o: Any = emptyArray(), tag: String? = null)  = handleMsg(WARN, tag, message, o)
    @JvmOverloads fun e(message: String?, vararg o: Any = emptyArray(), tag: String? = null)  = handleMsg(ERROR, tag, message, o)
    @JvmOverloads fun a(message: String?, vararg o: Any = emptyArray(), tag: String? = null)  = handleMsg(ASSERT, tag, message, o)
    @JvmOverloads fun v(@StringRes message: Int, vararg o: Any = emptyArray()) = handleMsg(VERBOSE,  null, str(message), o)
    @JvmOverloads fun d(@StringRes message: Int, vararg o: Any = emptyArray()) = handleMsg(DEBUG,    null, str(message), o)
    @JvmOverloads fun i(@StringRes message: Int, vararg o: Any = emptyArray()) = handleMsg(INFO,     null, str(message), o)
    @JvmOverloads fun w(@StringRes message: Int, vararg o: Any = emptyArray()) = handleMsg(WARN,     null, str(message), o)
    @JvmOverloads fun e(@StringRes message: Int, vararg o: Any = emptyArray()) = handleMsg(ERROR,    null, str(message), o)
    @JvmOverloads fun a(@StringRes message: Int, vararg o: Any = emptyArray()) = handleMsg(ASSERT,   null, str(message), o)
    @JvmOverloads fun ex (t: Throwable, message: String, vararg o: Any = emptyArray(), tag: String? = null) = handleExc(EX_FULL, t, tag, message, o)
    @JvmOverloads fun exs(t: Throwable, message: String, vararg o: Any = emptyArray(), tag: String? = null) = handleExc(EX_SIMPLER, t, tag, message, o)
    @JvmOverloads fun exm(t: Throwable, message: String, vararg o: Any = emptyArray(), tag: String? = null) = handleExc(EX_SIMPLEST, t, tag, message, o)
    @JvmOverloads fun ex (t: Throwable, @StringRes message: Int, vararg o: Any = emptyArray(), @StringRes tag: Int = 0) = handleExc(EX_FULL, t, str(tag), str(message), o)
    @JvmOverloads fun exs(t: Throwable, @StringRes message: Int, vararg o: Any = emptyArray(), @StringRes tag: Int = 0) = handleExc(EX_SIMPLER, t, str(tag), str(message), o)
    @JvmOverloads fun exm(t: Throwable, @StringRes message: Int, vararg o: Any = emptyArray(), @StringRes tag: Int = 0) = handleExc(EX_SIMPLEST, t, str(tag), str(message), o)
}

interface LogListener {
    /**
     * @return Return true, if you want to handle displaying logs
     */
    fun handleMsg(lvl: Int, tag: String?, msg: String?): Boolean

    /**
     * @return Return true, if you want to handle displaying logs
     */
    fun handleExc(simple_lvl: Int, t: Throwable, tag: String?, msg: String?, vararg o: Any): Boolean
}


internal class LoggerInternal {
    companion object {
        fun str(@StringRes id: Int) = if (id != 0) resources.getString(id) else null
        fun Throwable.printCause(tag: String?) {
            android.util.Log.w(tag, "Caused by: <${this.javaClass.simpleName}> ${this.message.toString()}")
        }

        fun printLocalCauseByStack(tag: String?, depth: Int, it: Throwable?): Unit = when {
            depth == 0 || it == null -> Unit
            else -> {
                it.printCause(tag)
                printLocalCauseByStack(tag, depth -1, it.cause)
            }
        }

        fun printLocalExceptionStack(tag: String?, pack: String, it: Throwable) {
            it.cause?.printCause(tag)
            it.stackTrace
                    ?.filter { !it.className.startsWith(pack) }
                    ?.forEach { android.util.Log.w(tag, it.toString()) }
        }
    }

}